Changing ipython directives to doctest/plot directives.

sarsid opened the pull request:
This pull request addresses #115, it changes almost all ipython directives to doctest directives.

nathanshammah commented:
Looking forward to this PR. This will close #114 and close also close issue #64.

sarsid commented:
I added a `.rst` file on how to write directives and use the doctest and nbplot. I am wondering where exactly we should put this instruction file. Any ideas? I know Jake is hosting the dev guide on his own website right now. Eventually I'd like this to be integrated into that guide. Otherwise, the PR is mostly ready. 

cc: @nathanshammah @BoxiLi

nathanshammah commented:
Thanks @sarsid for the push here. I had a look at `nb2plots` after our chat. It is not a very collaborative project and is not actively released on pypi, the last version is over two years old. I would recommend sticking to matplotlib's own directives (`plot`), using the standard sphinx extensions. My understanding is that this should work also for non-interactive python blocks, i.e., without ipython, is that correct?

Happy to chat about it if it helps.

nathanshammah commented:
I would also suggest splitting the PR in two: one just addressing the conversion between the ipython and doctest / code-block /testcode directives, the second one taking care of plots. 
https://matplotlib.org/sampledoc/extensions.html

BoxiLi commented:
From our chat and @sarsid 's writing, the advantages of using nb2plot are
- It supports `doctest` together with plotting. It is nice but I think the code output and test should be easy to separate? Does it also test if the plot is correctly generated?
- It uses the same namespace in one `rst` file. This is very nice since including all the headers in every code snippet might be very inconvenient e.g. in `guide-bloch.rst`, where we show things step by step.

Do you know if the above advantage has a replacement in matplotlib @sarsid ?

BTW, I successfully rendered the doc as a whole. It's a big step. Thanks @sarsid !

sarsid commented:
I think (with some tweaks) we can support both those features in the matplotlib plot directive as well. The only thing we will be missing is the useful nb2plot feature which converts notebooks to `.rst` files. I am changing the plot directive now. The only thing I would like is to not make a separate PR for the plots because it would mean I have to rework the code to ipython for the plots for now. I don't have a separate commit that neatly delineates the two so it would be quite a task. I'll keep this separation idea in mind from the next time for sure.

jakelishman commented:
The `.. plot` directive supports doctest syntax, and unified code context with the `:context` option: https://matplotlib.org/3.1.3/devel/plot_directive.html

> I know Jake is hosting the dev guide on his own website right now

It's actually just hosted in the `gh-pages` branch of [a repo in my account](https://github.com/jakelishman/qutip-devguide/), and it needs quite a lot of updating.  We were waiting til the docs built again before actually trying to merge it into a QuTiP repo.  If they work now, we can start thinking about it.

sarsid commented:
I have made all the changes from nbplot to plot directive. It works quite well now. I think the PR is ready for review. I would also recommend checking out the directives.rst file since that contains the guide on how to use doctest/plot. I don't know where to link it in the html directory yet. Let me know if you have any suggestions (that reference needs to be added in the readme as well).

BoxiLi commented on guide/guide-random.rst:
```diff
@@ -58,11 +74,43 @@ full-rank for the given dimension.
 For example,
 
-.. ipython::
+.. doctest:: [random]
+  :skipif: True
+
+   >>> rand_dm(5, density=0.5)
+   Quantum object: dims = [[5], [5]], shape = (5, 5), type = oper, isherm = True
```
In the case of testing a random generation, you could also try to set a random seed (in a hidden block) and then test it.

BoxiLi commented on guide/guide-states.rst:
```diff
@@ -267,59 +493,138 @@ Qubit (two-level) systems
+.. testoutput:: [states]
+    :options: +NORMALIZE_WHITESPACE
 
-    In [1]: vac
+    Quantum object: dims = [[2], [1]], shape = (2, 1), type = ket
+    Qobj data =
+    [[1.]
+     [0.]]
 
-.. ipython::
+.. testcode:: [states]
 
-    In [1]: c = create(2)
+    c = create(2)
 
-    In [2]: c * vac
+    print(c * vac)
 
+.. testoutput:: [states]
+    :options: +NORMALIZE_WHITESPACE
+
+    Quantum object: dims = [[2], [1]], shape = (2, 1), type = ket
+    Qobj data =
+    [[0.]
+     [1.]]
 
 For a spin system, the operator analogous to the raising operator is the sigma-plus operator :func:`qutip.operators.sigmap`.  Operating on the ``spin`` state gives:
 
-.. ipython::
+.. testcode:: [states]
+
+    print(spin)
+
+.. testoutput:: [states]
+    :options: +NORMALIZE_WHITESPACE
+
+    Quantum object: dims = [[2], [1]], shape = (2, 1), type = ket
+    Qobj data =
+    [[1.]
+     [0.]]
```
Here and in the following lines, is there a way to distinguish input and output in the rendered version?

sarsid commented on guide/guide-basics.rst:
```diff
@@ -45,28 +44,66 @@ The key difference between classical and quantum mechanics lies in the use of op
+.. testcode:: [basics]
+  :skipif: True
 
+    r = np.random.rand(4, 4)
+    print(Qobj(r))
+
+.. testoutput:: [basics]
+  :skipif: True
+
+    Quantum object: dims = [[4], [4]], shape = (4, 4), type = oper, isherm = False
+    Qobj data =
+    [[0.80891494 0.80278898 0.33983968 0.04215302]
+     [0.98991331 0.22610373 0.70347075 0.42928218]
+     [0.60515625 0.09007662 0.63876191 0.94787943]
+     [0.19734676 0.97222882 0.68423903 0.3205153 ]]
```
> From the rendered version, it is a bit hard to distinguish which lines are inputs and which are outputs. Is there a way to make it clearer?
Maybe I can try comments but not inherently.

> Just to verify if I understand it correctly, because it is random, this won't need a test, so your skipped it.
Yeah, maybe I can try random seed.

sarsid commented on guide/dynamics/dynamics-data.rst:
```diff
@@ -52,27 +52,41 @@ Accessing Result Data
-    expect = True
-    num_expect = 2, num_collapse = 2, ntraj = 500
+.. doctest::
+  :options: +SKIP
+
+  >>> print(result)
+  Result object with mcsolve data.
+  ---------------------------------
+  expect = True
+  num_expect = 2, num_collapse = 2, ntraj = 500
 
 The first line tells us that this data object was generated from the Monte Carlo solver ``mcsolve`` (discussed in :ref:`monte`).  The next line (not the ``---`` line of course) indicates that this object contains expectation value data.  Finally, the last line gives the number of expectation value and collapse operators used in the simulation, along with the number of Monte Carlo trajectories run ...
 
-Now we have all the information needed to analyze the simulation results. To access the data for the two expectation values one can do::
+Now we have all the information needed to analyze the simulation results.
+To access the data for the two expectation values one can do:
+
+
+.. testcode::
+  :skipif: True
+
+  expt0 = result.expect[0]
+  expt1 = result.expect[1]
```
None of the code here is supposed to be executed, they could be replaced with something like code-blocks, I kept it like that cause it renders nicer.

sarsid commented on guide/dynamics/dynamics-bloch-redfield.rst:
```diff
@@ -255,11 +290,12 @@ As we have already discussed, the Bloch-Redfield master equation requires transf
 
 
-.. ipython::
+.. plot::
```
> * I noticed that some blocks using `plot` have no plots. In the rendered file they appear with a `source code` at the end, which belongs to the plot directive. It breaks the flow of the text a bit. Then I realized that you probably just want to keep the same namespace. If there is no easy way around it, just ignore those comments of mine.


`plot` is used regardless of whether there needs to be a plot or not so that all the code is in the same namespace, we can work around it but don't need to, let me try to see if I can remove the `source code` thing at the end. 


> * If we are about to use `>>>` to indicate input code, it would be good to be consistent throughout the doc. I noticed that in `testcode` blocks you are writing the input without `>>>`. Is it possible also use `>>>`?

It isn't possible to use `>>>` with `testcode`. I decided between the `testcode` and `doctest` as when it was useful, i.e. `doctest` when multiple things are to be output in quick succession and `testcode` otherwise. 

> * We can also add these conventions to `directives.rst`, or other guides for doc contributors.

I'll add some of these suggestions to the doc.

sarsid replied:
Is it okay if none of the plot directives show the link to source code?

BoxiLi replied:
> Is it okay if none of the plot directives show the link to source code? There isn't an easy way to do it for individual blocks.

That would be excellent. If that is too much work you can also leave it. I get now that you want to keep the namespace.

sarsid commented on guide/dynamics/dynamics-floquet.rst:
```diff
@@ -104,96 +104,98 @@ Consider for example the case of a strongly driven two-level atom, described by
+   >>> q_energies = np.zeros((len(A_vec), 2))
+   >>> H0 = delta/2.0 * sigmaz() - eps0/2.0 * sigmax()
+   >>> args = {'w': omega}
+   >>> for idx, A in enumerate(A_vec): # doctest: +SKIP
+   >>>   H1 = A/2.0 * sigmax() # doctest: +SKIP
+   >>>   H = [H0, [H1, lambda t, args: sin(args['w']*t)]] # doctest: +SKIP
+   >>>   f_modes, f_energies = floquet_modes(H, T, args, True) # doctest: +SKIP
+   >>>   q_energies[idx,:] = f_energies # doctest: +SKIP
+   >>> figure() # doctest: +SKIP
+   >>> plot(A_vec/omega, q_energies[:,0] / delta, 'b', A_vec/omega, q_energies[:,1] / delta, 'r') # doctest: +SKIP
+   >>> xlabel(r'$A/\omega$') # doctest: +SKIP
+   >>> ylabel(r'Quasienergy / $\Delta$') # doctest: +SKIP
+   >>> title(r'Floquet quasienergies') # doctest: +SKIP
+   >>> show() # doctest: +SKIP
```
Is it appearing for you? It is not supposed to appear and isn't for me right now.

BoxiLi replied:
![image](https://user-images.githubusercontent.com/12125783/90321346-081dd180-df49-11ea-988d-18d58c7cc853.png)

It rendered like this for me. It is not for you? That's weird.

BTW, why skipping the test here?

sarsid replied:
Yeah, they arent rendering for me when in html mode atleast. 

I am skipping the tests because they seem to be glitching a bit (the plots are not being tested anyway given we dont have any meaninful output values for them). They will be "tested" when doing *make html* anyway.

sarsid commented on guide/dynamics/dynamics-floquet.rst:
```diff
@@ -104,96 +104,98 @@ Consider for example the case of a strongly driven two-level atom, described by
+
+   >>> T = 2*pi / omega
+   >>> f_modes_0, f_energies = floquet_modes(H, T, args)
+   >>> f_energies # doctest: +NORMALIZE_WHITESPACE
+   array([-2.83131212,  2.83131212])
+   >>> f_modes_0 # doctest: +NORMALIZE_WHITESPACE
```
Is it appearing for you? It is not supposed to appear and isn't for me right now.

sarsid commented on guide/dynamics/dynamics-floquet.rst:
```diff
@@ -104,96 +104,98 @@ Consider for example the case of a strongly driven two-level atom, described by
-
-   In [1]: psi_t
+   >>> t = 10 * np.random.rand()
+   >>> psi_t = floquet_wavefunction_t(f_modes_0, f_energies, f_coeff, t, H, T, args)
```
Do you mean it is incorrect on the website as well? 
http://qutip.org/docs/latest/guide/dynamics/dynamics-floquet.html

Right now, for me, it is showing the same as there.

sarsid commented:
Update:

1. added seeds to extended to random tests
2. fixed separation issue by adding * Output * before each output block
3. added more guidelines to `directives.rst`

Should be ready for review now.

sarsid commented:
> Very nice contributing guide!
> 
> With this PR, I can fully build the documentation locally. Although there are still a bunch of falling when I run doctest on my machine, most of them are float inaccuracy and white space problem. I think it is good to merge this as it is and deal with those failinglater.

I think those will be fixed when you merge the other doctest PR on the qutip repo.
